### ISSUE TITLE
Upgrade TUF from deprecated v0 API to modern v2

### ISSUE THREAD

**Issue by jas4711** [sigstore/sigstore#1871]
Hi!

I'm working on sigstore packaging in Debian, and due to other packages we would like to upgrade TUF (github.com/theupdateframework/go-tuf) from v0 to v2 and sigstore/sigstore is one of two packages still using the old interface. The v0.7 branch is deprecated, see:

https://github.com/theupdateframework/go-tuf#history---legacy-go-tuf-vs-go-tufv2

This is a issue to request that you update the code to TUF v2. Is this possible? Any reason not to?

**Comment by bobcallaway** [comment]
@jku @haydentherapper

**Comment by haydentherapper** [comment]
Hey, we have marked the Sigstore TUF client in this repo as deprecated. It has been rewritten using go-TUF v2 in the Sigstore-go repo, https://github.com/sigstore/sigstore-go/tree/main/pkg/tuf. While functionality is not 100% identical, the only significant change was dropping an unused custom metadata format, so you should be good to leverage the newer client.

**Comment by jas4711** [comment]
Thank you!  I will try to get sigstore/sigstore in Debian to build without the TUF dependency, assuming no lack of (desired) functionality.

**Comment by haydentherapper** [comment]
Are you trying to build Cosign? If so, it'll be a bit hard to pull out the dependency, but if not, it should be doable for other Sigstore Go projects.

**Comment by jas4711** [comment]
Yes, one ultimate goal is getting cosign to build, but current project on the way to that bigger goal is getting sigstore-go to build.  I'll attempt to disable TUF for sigstore-sigstore and worry about cosign later.

**Comment by haydentherapper** [comment]
For Cosign, I'll mention that we are working on integrating the new sigstore-go TUF client into the current version. We'll still have go-tuf v0.7 in the dependency tree though, until the next major version of Cosign when we drop support for the old Sigstore TUF client.

Just wondering, is there a particular reason you need to remove go-tuf v0.7? It's stable, albeit no longer updated, and should be free of vulns.

**Comment by jas4711** [comment]
Okay.

The problem is that supporting two concurrent incompatible APIs for the same namespace is not possible in Debian unless upstream clearly separate things in different v0/ or v2/ sub-directories.  So there is no simple way to support both TUF v0 and v2 in the same dependency path.  And sigstore-go pulls in both v2 directly and v0 indirectly...   As comparison, the golang jwt project seems to have well-behaved APIs that are in different namespaces, so both packages can be installed at the same time.  But TUF v0 and v2 seems to collide in their APIs.

Sorry if this is somewhat confused, I'm not a Go person, and the way Go code is packaged in Debian is somewhat strange to Go people so there is some significant impedance mismatch here.  Thanks for all your efforts!

**Comment by haydentherapper** [comment]
Lemme ask around, as I'm also not very familiar with these details in Go namespacing.

**Comment by codysoyland** [comment]
@jas4711 Is the project you're building with using Go Modules or some other method of collecting dependencies like Bazel? Go Modules have been the standard in the Go community for many years now (replacing Dep, Glide, and others), and it natively supports concurrently importing multiple major versions of dependencies even if they share the same directory structure with conflicting names.

[This document](https://go.dev/wiki/Modules#releasing-modules-v2-or-higher) explains how major versions are released, and I wonder if your tooling is expecting the `v2` branch in the go-tuf dependency, which currently doesn't exist. I did notice that jwt uses the major branch strategy and [includes `v3`, `v4`, and `v5` branches](https://github.com/golang-jwt/jwt/branches/all), so that may be the distinction that allows that dependency to work for you. We can ask the go-tuf maintainers to do the same, but there may be other options for you. Have you tried using `go mod vendor`? It will gather the major versions and put them in separate directories for you in the `vendor` subdirectory. Perhaps this is an option you can use?

**Comment by kommendorkapten** [comment]
Copied [this message](https://sigstore.slack.com/archives/C03SZ6SHU90/p1730186994054689?thread_ts=1729895186.390399&cid=C03SZ6SHU90)

Looking at `golang-jwt`, although they have different branches like `v5`, they are releasing from `main`, e.g. `v5.2.1` is released from `main`, the branch `v5` is outdated.

See here: 
* https://github.com/golang-jwt/jwt/tree/v5
* https://github.com/golang-jwt/jwt/releases/tag/v5.2.1
* https://github.com/golang-jwt/jwt/commits/main/

The commit that matches `v5.2.1` is from February 12th.

**Comment by kommendorkapten** [comment]
@jas4711 I would be interested to see what version you are getting from `golang-jwt` when you are building, as they the `v5` branch is behind the latest `v5.2.1` release.

**Comment by jas4711** [comment]
Hi again.  I wonder if this issue really should be closed.  github.com/sigstore/cosign uses github.com/sigstore/sigstore/pkg/tuf which depends on TUF v0.  While it is okay if you consider this deprecated, it doesn't really change anything since it is still shipped in the release and used by other projects.  Do you plan to make a new release without pkg/tuf v0 dependency?  Once that is done I think this could be closed, or is that a separate issue?

Sorry if I'm missing how dependencies work here, I'm packaging `cosign` for Debian and their build systems is rather different to what I expect you as upstream Go people are familiar with.

**Comment by jas4711** [comment]
> @jas4711 I would be interested to see what version you are getting from `golang-jwt` when you are building, as they the `v5` branch is behind the latest `v5.2.1` release.

Debian is at golang-jwt v5.2.1, it tracks the tags.  https://tracker.debian.org/pkg/golang-github-golang-jwt-jwt-v5
